### DISCUSSION TITLE
Invoking in a serverless context

### DISCUSSION THREAD

**Discussion by drewpearce** [spec-first/connexion#1827] [Q&A] [answered]
### Context

We utilize connexion (v2) in AWS Lambda currently, and since there's no actual python server software running, we wrote our own handler that takes the lambda invocation event, creates the connection app (`connextion.FlaskApp`) from the spec, and then gets the API response via `werkzeug.wrappers.Response.from_app`. I'm in the process of migrating this tooling to connexion v3 and this way of invocation no longer works, as it passes the wrong arguments to the app. It looks like this is because the Werkzeug wrapper assumes a WSGI app.

### My Question

Is there a recommended way to do this sort of setup in connexion v3 (send request and get a response directly without the need for a server wrapper)?

**Comment by RobbeSneyders** [comment, marked as answer]
Hi @drewpearce,

Connexion now implements the ASGI interface, which is a Python interface that can be called directly. You'll probably need some boilerplate to encode your invocation event into the ASGI scope and channels though. This is what `werkzeug.wrappers.Response.from_app` provides for WSGI, but I'm not aware of any equivalent functionality for ASGI.

I think the most logical place to add this in the ASGI ecosystem, would be in [Starlette](https://github.com/encode/starlette/tree/master), or otherwise [asgiref](https://github.com/django/asgiref/). Starlette already offers most of the needed functionality as part of its [`TestClient`](https://github.com/encode/starlette/blob/master/starlette/testclient.py), but not with the interface you're looking for.

For a long term fix, I would propose that you open a discussion on the Starlette repo to check if this is something they would be open to support.

For a short term fix, I see two options:
- Use the `TestClient` interface ([docs](https://connexion.readthedocs.io/en/latest/testing.html)). Note that `werkzeug.wrappers.Response.from_app` also leverages testing functionality underneath.
- Wrap your Connexion application in an [`a2wsgi.ASGIMiddleware`](https://github.com/abersheeran/a2wsgi/tree/master#asgimiddleware) and use it with `werkzeug.wrappers.Response.from_app`. Note that this might increase the response time.

**Reply by drewpearce** [reply]
Thanks for the tips! Long-term we will definitely do an overhaul.
